File: src/nl/pindab0ter/aoc2023/day01/Trebuchet.kt
@file:Suppress("SpellCheckingInspection")

package nl.pindab0ter.aoc2023.day01

fun main() {
    val input = ClassLoader.getSystemResource("2023/day01/input").readText()
    val result = input.lines().sumOf(::getCalibrationValue)

    println(result)
}

val digitNames = mapOf(
    "one" to 1,
    "two" to 2,
    "three" to 3,
    "four" to 4,
    "five" to 5,
    "six" to 6,
    "seven" to 7,
    "eight" to 8,
    "nine" to 9,
)

fun getCalibrationValue(line: String): Int {
    val digitRegex = Regex("""(one|two|three|four|five|six|seven|eight|nine|\d)""")
    val matchFirst = digitRegex.find(line)!!.value

    // Reversed digit regex to cover edge cases like "eightwo", as Regex doesn't support overlapping matches.
    // Thanks, Bart
    val reversedDigitRegex = Regex("""(eno|owt|eerht|ruof|evif|xis|neves|thgie|enin|\d)""")
    val matchLast = reversedDigitRegex.find(line.reversed())?.value?.reversed()

    val firstDigit = digitNames[matchFirst] ?: matchFirst
    val lastDigit = when (matchLast) {
        null -> firstDigit
        else -> digitNames[matchLast] ?: matchLast
    }

    return "${firstDigit}${lastDigit}".toInt()
}

File: src/main/clojure/nl/pindab0ter/aoc2023/day01/trebuchet.clj
(ns nl.pindab0ter.aoc2023.day01.trebuchet
  (:require
    [clojure.string :as str]
    [nl.pindab0ter.common.advent-of-code :refer [get-input]]
    [nl.pindab0ter.common.collections :refer [sum]]))


(defn get-digit [s]
  (let [digit (re-find #"^\d" s)]
    (cond
      (not (nil? digit)) (Integer/parseInt digit)
      (str/starts-with? s "one") 1
      (str/starts-with? s "two") 2
      (str/starts-with? s "three") 3
      (str/starts-with? s "four") 4
      (str/starts-with? s "five") 5
      (str/starts-with? s "six") 6
      (str/starts-with? s "seven") 7
      (str/starts-with? s "eight") 8
      (str/starts-with? s "nine") 9)))


(defn get-calibration-value
  "Get the calibration value from a line."
  ([s] (get-calibration-value s []))
  ([s coll]
   (if (empty? s)
     (let [digits (filter identity coll)]
       (Integer/parseInt (str (first digits) (last digits))))
     (let [digit (get-digit s)]
       (recur (subs s 1) (conj coll digit))))))


(defn -main []
  (let [input  (get-input 2023 1)
        values (map get-calibration-value (str/split-lines input))]
    (println (sum values))))
